jetty-http/src/main/java/org/eclipse/jetty/http/compression/Huffman.java

```diff
@@ -348,199 +347,46 @@
-
-                    // terminal node
-                    utf8.append((byte)(0xFF & rowsym[node]));
-                    bits -= rowbits[node];
-                    node = 0;
-                }
-                else
-                {
-                    // non-terminal node
-                    bits -= 8;
-                }
-            }
-        }
-
-        while (bits > 0)
-        {
-            int c = (current << (8 - bits)) & 0xFF;
-            int lastNode = node;
-            node = tree[node * 256 + c];
-
-            if (rowbits[node] == 0 || rowbits[node] > bits)
-            {
-                int requiredPadding = 0;
-                for (int i = 0; i < bits; i++)
-                {
-                    requiredPadding = (requiredPadding << 1) | 1;
-                }
-
-                if ((c >> (8 - bits)) != requiredPadding)
-                    throw new HpackException.CompressionException("Incorrect padding");
-
-                node = lastNode;
-                break;
-            }
-
-            utf8.append((byte)(0xFF & rowsym[node]));
-            bits -= rowbits[node];
-            node = 0;
-        }
-
-        if (node != 0)
-            throw new HpackException.CompressionException("Bad termination");
-
-        return utf8.toString();
-    }
-
-    public static int octetsNeeded(String s)
-    {
-        return octetsNeeded(CODES, s);
-    }
-
-    public static int octetsNeeded(byte[] b)
-    {
-        return octetsNeeded(CODES, b);
-    }
-
-    public static void encode(ByteBuffer buffer, String s)
-    {
-        encode(CODES, buffer, s);
-    }
-
-    public static void encode(ByteBuffer buffer, byte[] b)
-    {
-        encode(CODES, buffer, b);
-    }
-
-    public static int octetsNeededLC(String s)
-    {
-        return octetsNeeded(LCCODES, s);
-    }
-
-    public static void encodeLC(ByteBuffer buffer, String s)
-    {
-        encode(LCCODES, buffer, s);
-    }
-
-    private static int octetsNeeded(final int[][] table, String s)
-    {
-        int needed = 0;
-        int len = s.length();
-        for (int i = 0; i < len; i++)
-        {
-            char c = s.charAt(i);
-            if (c >= 128 || c < ' ')
-                return -1;
-            needed += table[c][1];
-        }
-
-        return (needed + 7) / 8;
-    }
-
-    private static int octetsNeeded(final int[][] table, byte[] b)
+    public static boolean isIllegalCharacter(char c)
     {
-        int needed = 0;
-        int len = b.length;
-        for (int i = 0; i < len; i++)
-        {
-            int c = 0xFF & b[i];
-            needed += table[c][1];
-        }
-        return (needed + 7) / 8;
+        return (c >= 256 || c < ' ');
     }
 
-    /**
-     * @param table The table to encode by
-     * @param buffer The buffer to encode to
-     * @param s The string to encode
-     */
-    private static void encode(final int[][] table, ByteBuffer buffer, String s)
+    public static char getReplacementCharacter(char c)
```

Comment:
protected?

also it needs a bit of javadoc and links to why replacement character is different.

plus it needs to be renamed, as it is sanitising not replacing.  Probably also should be on HttpToken